go/vt/vtgate/semantics/transitive_closures.go

```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2025 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package semantics
+
+import (
+	"vitess.io/vitess/go/vt/sqlparser"
+)
+
+type (
+	// TransitiveClosures expression equality, so we can rewrite expressions efficiently
+	// It allows for transitive closures (e.g., if a == b and b == c, then a == c).
+	TransitiveClosures struct {
+		// ExprIDs stores the IDs of the expressions.
+		// The ID can then be used to get other equivalent expressions from the Equalities slice
+		exprIDs map[sqlparser.Expr]int
```

Comment:
We discussed a more optimal implementation for this using Disjoin set unions - https://en.wikipedia.org/wiki/Disjoint-set_data_structure. Left that to be done later, just pointing out so that we have the reference to it.